File: partiql-plan/src/main/java/org/partiql/plan/rex/RexNullIf.java
package org.partiql.plan.rex;

import org.jetbrains.annotations.NotNull;
import org.partiql.plan.Operand;
import org.partiql.plan.OperatorVisitor;

import java.util.List;

/**
 * Logical nullif expression abstraction base class.
 */
public abstract class RexNullIf extends RexBase {

    /**
     * @return new RexNullIf instance
     */
    @NotNull
    public static RexNullIf create(@NotNull Rex v1, @NotNull Rex v2) {
        return new Impl(v1, v2);
    }

    /**
     * @return v1 rex (operand 0)
     */
    @NotNull
    public abstract Rex getV1();

    /**
     * @return v2 rex (operand 1)
     */
    @NotNull
    public abstract Rex getV2();

    /**
     * @return minimal common supertype of (NULL, typeof(v1))
     */
    @NotNull
    @Override
    protected final RexType type() {

        return getV1().getType();
    }

    @NotNull
    @Override
    protected final List<Operand> operands() {
        Operand c0 = Operand.single(getV1());
        Operand c1 = Operand.single(getV2());
        return List.of(c0, c1);
    }

    @Override
    public <R, C> R accept(OperatorVisitor<R, C> visitor, C ctx) {
        return visitor.visitNullIf(this, ctx);
    }

    private static class Impl extends RexNullIf {

        private final Rex v1;
        private final Rex v2;

        private Impl(Rex v1, Rex v2) {
            this.v1 = v1;
            this.v2 = v2;
        }

        @NotNull
        @Override
        public Rex getV1() {
            return v1;
        }

        @NotNull
        @Override
        public Rex getV2() {
            return v2;
        }
    }
}

File: partiql-plan/src/main/kotlin/org/partiql/plan/rex/RexNullIf.kt
package org.partiql.plan.rex

/**
 * Logical operator for the SQL NULLIF special form.
 */
public interface RexNullIf : Rex {

    public fun getV1(): Rex

    public fun getV2(): Rex

    override fun getChildren(): Collection<Rex> = listOf(getV1(), getV2())

    override fun <R, C> accept(visitor: RexVisitor<R, C>, ctx: C): R = visitor.visitNullIf(this, ctx)
}

/**
 * Internal
 */
internal class RexNullIfImpl(v1: Rex, v2: Rex) : RexNullIf {

    // DO NOT USE FINAL
    private var _v1 = v1
    private var _v2 = v2

    override fun getV1(): Rex = _v1

    override fun getV2(): Rex = _v2

    override fun getType(): RexType = _v1.getType()

    override fun getChildren(): Collection<Rex> = listOf(_v2)

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is RexNullIf) return false
        if (_v1 != other.getV1()) return false
        if (_v2 != other.getV2()) return false
        return true
    }

    override fun hashCode(): Int {
        var result = _v1.hashCode()
        result = 31 * result + _v2.hashCode()
        return result
    }
}
